core/src/test/java/org/infinispan/persistence/PersistenceManagerTest.java

```diff
@@ -53,26 +55,32 @@ public void testStopDuringProcess() throws ExecutionException, InterruptedExcept
       PersistenceManager persistenceManager = extractComponent(cache, PersistenceManager.class);
       KeyPartitioner keyPartitioner = extractComponent(cache, KeyPartitioner.class);
       //simulates the scenario where, concurrently, the cache is stopped during a process loop
-      persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k1", "v1", cache), keyPartitioner.getSegment("k1"), BOTH);
-      persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k2", "v2", cache), keyPartitioner.getSegment("k2"), BOTH);
-      persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k3", "v3", cache), keyPartitioner.getSegment("k3"), BOTH);
+      CompletionStages.join(persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k1", "v1", cache), keyPartitioner.getSegment("k1"), BOTH));
+      CompletionStages.join(persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k2", "v2", cache), keyPartitioner.getSegment("k2"), BOTH));
+      CompletionStages.join(persistenceManager.writeToAllNonTxStores(MarshalledEntryUtil.create("k3", "v3", cache), keyPartitioner.getSegment("k3"), BOTH));
       final CountDownLatch before = new CountDownLatch(1);
       final CountDownLatch after = new CountDownLatch(1);
-      final AtomicInteger count = new AtomicInteger(0);
-      Future<Object> c = fork(() -> Flowable.fromPublisher(persistenceManager.publishEntries(true, true))
-            .subscribe(ignore -> {
-               before.countDown();
-               after.await();
-               count.incrementAndGet();
-            }));
-      before.await(30, TimeUnit.SECONDS);
+      Future<Integer> c = fork(() -> {
+         TestSubscriber<Object> subscriber = TestSubscriber.create(0);
+         Flowable.fromPublisher(persistenceManager.publishEntries(true, true))
+               .subscribe(subscriber);
+         before.countDown();
+         assertTrue(after.await(10, TimeUnit.SECONDS));
+         // request all the elements after we have initiated stop (3 elements with 100ms wait for each will be run)
+         subscriber.request(Long.MAX_VALUE);
+         subscriber.await(10, TimeUnit.SECONDS);
+         subscriber.assertNoErrors();
+         subscriber.assertComplete();
+         return subscriber.valueCount();
+      });
+      assertTrue(before.await(30, TimeUnit.SECONDS));
       Future<Void> stopFuture = fork(persistenceManager::stop);
-      //stop is unable to proceed while the process isn't finish.
-      expectException(TimeoutException.class, () -> stopFuture.get(1, TimeUnit.SECONDS));
+      //stop is unable to proceed while the process isn't finish - note that with slow store the publisher should take 300+ ms
+      expectException(TimeoutException.class, () -> stopFuture.get(150, TimeUnit.MILLISECONDS));
```

Comment:
This looks risky, the main thread could easily start `stopFuture.get()` 200ms after `persistenceManager.stop()` when running in parallel with other tests. I suggest blocking the publisher with a latch instead.

In fact I would really like to kill the `slowStore` config, I feel it slows things too much because you can't choose which operation to slow down, and 100ms is either too much or not enough.

Reply:
Unfortunately the way the store config is setup, there is no way to pass an instance to use a latch, otherwise I would have. I can reduce it to 50 ms instead to make it less likely.

Reply:
I don't think you need to pass an instance to the config (although you could use `CacheStoreFactoryRegistry` `clearFactories()` + `addCacheStoreFactory()` to do even that). You only need to put a latch or a `CheckPoint` in a `DummyInMemoryStoreConfiguration` attribute with an `IdentityAttributeCopier`. Or, if you don't need it during startup, as is the case here, you can add a setter to `DummyInMemoryStore` and set it after `PersistenceManager` created it.

That being said, maybe it's still too much to add to this PR, so I'm ok with the lower timeout for now.

Reply:
Yeah, I know there are ways to do it, but it is really convoluted and difficult to understand what is going on. The whole persistence stuff should be revamped a bit to be better for testing imo.